### ISSUE TITLE
Optimize remote calls for large wrapper jobs

### ISSUE THREAD

**Issue by kinow** [BSC-ES/autosubmit#329]
In GitLab by @ lbatista on Apr 10, 2018, 18:53

@mcastril @dmanubens I decided to open an issue about this because it allows us to better trace it in the future, but also because I was wondering what could be the possible implications.
As we have discussed, for a wrapper job consisting of ~300 jobs, it took 2h30 just repeating the same mkdir (to create the remote LOG directory which already existed) and removing STAT and COMPLETED files. 

My idea was to change the `_do_submission` method of the `JobPackageThread` class (the one inherited by the wrapper packages) as follows:

    def _do_submission(self):
        filenames = str()
        for job in self.jobs:
            filenames += " " + job.name + "_STAT " + job.name + "_COMPLETED"
        self.platform.remove_multiple_files(filenames)
        
        package_id = self.platform.submit_job(None, self._common_script)

        if package_id is None:
            raise Exception('Submission failed')

        for i in range(1, len(self.jobs) + 1):
            Log.info("{0} submitted", self.jobs[i - 1].name)
            self.jobs[i - 1].id = str(package_id)
            self.jobs[i - 1].status = Status.SUBMITTED
            self.jobs[i - 1].write_submit_time()

This `remove_multiple_files` I added to `ParamikoPlatform` class:

    def remove_multiple_files(self, filenames):
        command = "rm " + filenames

        if self.send_command(command):
            return self._ssh_output
        else:
            return None

As for the mkdirs, it happens inside the send_file method that sends a cmd to the platform. First it checks the remote directory, removes previously existing cmd, and sends the current cmd. In this case the idea would be to check the remote directory first, then delete all the cmds in a single command using the `remove_multiple_files` and then sending the current cmds one by one.

Do you see any pitfalls or drawbacks in this approach? It also relies less on Paramiko's API, as it's basically sending raw commands (with a parser for the output if necessary)... I wanted to get your feedback before going ahead with this

**Comment by kinow** [comment]
In GitLab by @mcastril on May 4, 2018, 18:06

Sorry @ lbatista for arriving so late to this. I will be try to be more close to AS issues in the near term, now that some deadlines have been surpassed.

In any case I assume that @dmanubens gave you a feedback on this, he has much more knowledge on Autosubmit code. In any case if you put the issue hash in the commit's description, the commit can be accessed from here to see the fix. 

:)

**Comment by kinow** [comment]
In GitLab by @dmanubens on Jul 19, 2018, 15:47

I could see pitfall when using ECMWF CCA rm and mkdir commands through ecaccess client

**Comment by kinow** [comment]
In GitLab by @ lbatista on Jul 25, 2018, 17:35

@mcastril @dmanubens I have been trying to run an experiment (dummy one for now) to test the wrapper on CCA and I am getting the error below. Is it related to this change?

```
6 of 6 jobs remaining (17:25)
Jobs ready for ecmwf-cca: 1
/bin/sh: ecaccess-file-mkdir: command not found
/bin/sh: ecaccess-file-mkdir: command not found
[ERROR] Could not execute command ecaccess-file-mkdir cca:/scratch/ms/spesiccf/c3d/a01y; ecaccess-file-mkdir cca:/scratch/ms/spesiccf/c3d/a01y/LOG_a01y on cca
[ERROR] Could not create the DIR /scratch/ms/spesiccf/c3d/a01y/LOG_a01y on HPC cca
[ERROR] Could not execute command rm  /scratch/ms/spesiccf/c3d/a01y/LOG_a01y/a01y_20000101_fc0_1_SIM.cmd /scratch/ms/spesiccf/c3d/a01y/LOG_a01y/a01y_20000101_fc0_2_SIM.cmd /scratch/ms/spesiccf/c3d/a01y/LOG_a01y/a01y_20000101_fc0_3_SIM.cmd on cca
rm: cannot remove '/scratch/ms/spesiccf/c3d/a01y/LOG_a01y/a01y_20000101_fc0_1_SIM.cmd': No such file or directory
rm: cannot remove '/scratch/ms/spesiccf/c3d/a01y/LOG_a01y/a01y_20000101_fc0_2_SIM.cmd': No such file or directory
rm: cannot remove '/scratch/ms/spesiccf/c3d/a01y/LOG_a01y/a01y_20000101_fc0_3_SIM.cmd': No such file or directory
/bin/sh: ecaccess-file-mkdir: command not found
/bin/sh: ecaccess-file-mkdir: command not found
[ERROR] Could not execute command ecaccess-file-mkdir cca:/scratch/ms/spesiccf/c3d/a01y; ecaccess-file-mkdir cca:/scratch/ms/spesiccf/c3d/a01y/LOG_a01y on cca
[ERROR] Could not create the DIR /scratch/ms/spesiccf/c3d/a01y/LOG_a01y on HPC cca
/bin/sh: ecaccess-file-delete: command not found
/bin/sh: ecaccess-file-put: command not found
[ERROR] Could not send file /home/Earth/lbatista/autosubmit/a01y/tmp/a01y_20000101_fc0_1_SIM.cmd to /scratch/ms/spesiccf/c3d/a01y/LOG_a01y/a01y_20000101_fc0_1_SIM.cmd
[ERROR] ecmwf-cca submission failed
[CRITICAL] Unhandled exception on Autosubmit: Command 'ecaccess-file-put /home/Earth/lbatista/autosubmit/a01y/tmp/a01y_20000101_fc0_1_SIM.cmd cca:/scratch/ms/spesiccf/c3d/a01y/LOG_a01y/a01y_20000101_fc0_1_SIM.cmd' returned non-zero exit status 127
Traceback (most recent call last):
  File "/home/Earth/lbatista/PycharmProjects/autosubmit/autosubmit/autosubmit.py", line 374, in parse_args
    return Autosubmit.run_experiment(args.expid, args.notransitive)
  File "/home/Earth/lbatista/PycharmProjects/autosubmit/autosubmit/autosubmit.py", line 785, in run_experiment
    if Autosubmit.submit_ready_jobs(as_conf, job_list, platforms_to_test, packages_persistence):
  File "/home/Earth/lbatista/PycharmProjects/autosubmit/autosubmit/autosubmit.py", line 828, in submit_ready_jobs
    package.submit(as_conf, job_list.parameters)
  File "/home/Earth/lbatista/PycharmProjects/autosubmit/autosubmit/job/job_packages.py", line 87, in submit
    self._send_files()
  File "/home/Earth/lbatista/PycharmProjects/autosubmit/autosubmit/job/job_packages.py", line 293, in _send_files
    self.platform.send_file(self._job_scripts[job.name], check=False)
  File "/home/Earth/lbatista/PycharmProjects/autosubmit/autosubmit/platforms/ecplatform.py", line 141, in send_file
    subprocess.check_call(command, shell=True)
  File "/usr/lib64/python2.7/subprocess.py", line 186, in check_call
    raise CalledProcessError(retcode, cmd)
CalledProcessError: Command 'ecaccess-file-put /home/Earth/lbatista/autosubmit/a01y/tmp/a01y_20000101_fc0_1_SIM.cmd cca:/scratch/ms/spesiccf/c3d/a01y/LOG_a01y/a01y_20000101_fc0_1_SIM.cmd' returned non-zero exit status 127


Process finished with exit code 1
```

**Comment by kinow** [comment]
In GitLab by @mcastril on Jul 31, 2018, 11:19

Hi @ lbatista , was this related to not using the ECAC module? Was it solved?

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on May 17, 2019, 15:08

Hi,

@mcastril with the changes performed in 3.12.1b I think that this is not necessary anymore so, I will close this issue,

I really forgot about this issue but, I changed the way of how autosubmit delete files and creates in 3.12.1b because, until now it was using send_command (which creates a new channel and close it each time) two times per each time that autosubmit wanted to delete or create a single file, one for check if it exists and another to delete or create it.


Now for all platforms, the deletion is made with SFTP and regarding to creation, it looks first with sftp if it exists and if not, it creates the folder/file via send_command (SFTP has the limitation of being unable to perform a recursive creation).

The reason for that is that SFTP is now much faster (almost instant) because it has been changed for keeping the session open all runtime.
